File: ruby/bob/bob.rb
class Bob

  def hey(message)
    answer(MessageType.new(message))
  end

  private

    def answer(message_type)
      case 
      when message_type.silent?
        annoyed
      when message_type.yell?
        anxious
      when message_type.question? 
        approve
      else
        neutral
      end
    end

    def approve
      "Sure."
    end

    def annoyed
      "Fine. Be that way."
    end

    def anxious
      "Woah, chill out!"
    end

    def neutral
      "Whatever." 
    end

end

class MessageType
  
  attr_reader :message
  def initialize(message)
    @message = message.to_s
  end  

  def silent?
    message.strip.empty?
  end

  def question?
    message.end_with?("?")
  end

  def yell?
    message.upcase == message
  end

end

File: elixir/bob/bob.exs
defmodule Teenager do
  defdelegate yelling?(statement),  to: SentenceType
  defdelegate silence?(statement),  to: SentenceType
  defdelegate question?(statement), to: SentenceType

  def hey(statement) do
    cond do
      silence?(statement)  -> answer_for_silence
      yelling?(statement)  -> frightened_answer
      question?(statement) -> question_answer
      true                 -> default_answer
    end
  end

  defp frightened_answer do
    "Woah, chill out!"
  end

  defp question_answer do
    "Sure."
  end

  defp answer_for_silence do
    "Fine. Be that way."
  end

  defp default_answer do
    "Whatever."
  end
end

defmodule SentenceType do
  def yelling?(statement) do
    statement |> String.upcase == statement
  end

  def question?(statement) do
    statement |> String.ends_with?("?")
  end

  def silence?(statement) do
    statement == ""
  end
end
